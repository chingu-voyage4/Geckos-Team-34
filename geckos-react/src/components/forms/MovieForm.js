import React, { Component } from 'react';
import{ Form, Button, Rating } from 'semantic-ui-react';

import validator from 'validator';
import InlineError from '../messages/InlineError';
import ImageUpload from '../ImageUpload';
import PropTypes from 'prop-types';

class MovieForm extends Component {
  state = {
    title: '',
    director: '',
    plot: '',
    language: '',
    movieRating: '',
    rating: '',
    genre: '',
    releaseDate: '',
    producers: '',
    runTime: '',
    type: '',
    errors: {},
    loading: false
  }

  static propTypes = {
    submit: PropTypes.func.isRequired
  }

  validate = () => {
    const errors = {};
    if(!validator.isLength( this.state.title, { min: 2, max: 32 })) errors.title = 'Invalid Movie Title';
    if(!validator.isLength( this.state.director, { min: 2, max: 32 })) errors.director = 'Invalid Director Name';
    if(!validator.isLength( this.state.producers, { min: 2, max: 32 })) errors.producers = 'Invalid Producer Name';
    if(!validator.isLength( this.state.genre, { min: 2, max: 32 })) errors.genre = 'Invalid Genre Type';
    if(validator.isEmpty(this.state.plot)) errors.plot = 'Hey, you need to add a plot';
    if(!validator.isNumeric(this.state.runTime)) errors.runTime = 'Numbers only please';
    if(!validator.isNumeric(this.state.releaseDate)) errors.releaseDate = 'Numbers only please';

    return errors;
  }

  onInputChange = (e) => {
    const elem = e.target;
    const stateObj = {};

    stateObj[elem.name] = elem.value;
    this.setState(stateObj);
  }

  onRateChange = (e, { rating }) =>  {
    const stateObj = { rating };
    this.setState(stateObj);
  }

  onTypeChange = (e, { value }) => {
    const stateObj = { type: value };
    this.setState(stateObj);
  }

  onSubmit = () => {
    const errors = this.validate(this.state);
    this.setState(() => ({ errors }));

    if (Object.keys(errors).length === 0) {
      this.setState(() => ({ loading : true }));
      this.props.submit(this.state);
    }
  }

  getImgFile = (imgFile) => {
    const stateObj = { uploadedImage: imgFile };
    this.setState(stateObj);
  }

  render() {
    const { title, plot, language, movieRating, director, releaseDate,
      genre, producers, runTime, loading, errors } = this.state;

    return (
      <Form  loading={loading} onSubmit={this.onSubmit}>
        <Form.Group >
          <Form.Field width={4} error={!!errors.title}>
            <label htmlFor="title">Title</label>
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Mean Girls"
              onChange={this.onInputChange}
              value={title}
            />
            {
              errors.title && <InlineError message={errors.title} />
            }
          </Form.Field>
          <Form.Field width={2} error={!!errors.releaseDate}>
            <label htmlFor="releaseDate">Release Date</label>
            <input
              placeholder="2018"
              type="text"
              maxLength={4}
              name="releaseDate"
              id="releaseDate"
              onChange={this.onInputChange}
              value={releaseDate}
            />
            {
              errors.releaseDate && <InlineError message={errors.releaseDate} />
            }
          </Form.Field>
          <Form.Field width={2} error={!!errors.runTime}>
            <label htmlFor="runTime">Run Time(mins)</label>
            <input
              placeholder="120"
              type="text"
              maxLength={3}
              name="runTime"
              id="runTime"
              onChange={this.onInputChange}
              value={runTime}
            />
            {
              errors.runTime && <InlineError message={errors.runTime} />
            }
          </Form.Field>
        </Form.Group>
        <Form.Group inline>
          <label>Type</label>
          <Form.Radio label='Movie'
            name="type"
            value="movie"
            checked={this.state.type === "movie"}
            onChange={this.onTypeChange}
          />
          <Form.Radio label='TV'
            name="type"
            value="tv"
            checked={this.state.type === "tv"}
            onChange={this.onTypeChange}
          />
        </Form.Group>
        <Form.Group inline>
          <label htmlFor="rating">Your rating</label>
          <Rating
            defaultRating={1}
            maxRating={5}
            icon="star"
            name="rating"
            id="rating"
            onRate={this.onRateChange}
          >
          </Rating>
        </Form.Group>
        <Form.Group>
          <Form.Field width={2} error={!!errors.genre}>
            <label htmlFor="genre">Genre</label>
            <input
              type="text"
              id="genre"
              name="genre"
              placeholder="Action"
              onChange={this.onInputChange}
              value={genre}
            />
            {
              errors.genre && <InlineError message={errors.genre} />
            }
          </Form.Field>
          <Form.Field width={4} error={!!errors.director}>
            <label htmlFor="director">Director</label>
            <input
              type="text"
              id="director"
              name="director"
              placeholder="Some Director"
              onChange={this.onInputChange}
              value={director}
            />
            {
              errors.director && <InlineError message={errors.director} />
            }
          </Form.Field>
          <Form.Field width={4} error={!!errors.producers}>
            <label htmlFor="producers">Producers</label>
            <input
              type="text"
              id="producers"
              name="producers"
              placeholder="Some producer"
              onChange={this.onInputChange}
              value={producers}
            />
            {
              errors.producers && <InlineError message={errors.producers} />
            }
          </Form.Field>
        </Form.Group>
        <Form.Field width={8} error={!!errors.plot}>
          <label htmlFor="plot">Plot</label>
          <textarea
            name="plot"
            id="plot"
            cols="30"
            rows="10"
            onChange={this.onInputChange}
            value={plot}
          >
          </textarea>
          {
            errors.plot && <InlineError message={errors.plot} />
          }
        </Form.Field>
        <Form.Group>
          <Form.Field width={2}>
            <label htmlFor="language">Language</label>
            <select
              name="language"
              id="language"
              onChange={this.onInputChange}
              value={language}
            >
              <option value="english">English</option>
              <option value="spanish">Spanish</option>
              <option value="french">French</option>
              <option value="chinese">Chinese</option>
            </select>
          </Form.Field>
          <Form.Field width={2}>
            <label htmlFor="movieRating">Rating</label>
            <select
              name="movieRating"
              id="movieRating"
              onChange={this.onInputChange}
              value={movieRating}
            >
              <option value="g">G</option>
              <option value="pg">PG</option>
              <option value="pg-13">PG-13</option>
              <option value="r">R</option>
            </select>
          </Form.Field>
        </Form.Group>
        <ImageUpload imgFile={this.getImgFile} />
        <Button primary>Add Movie</Button>
      </Form>
    );
  }
};

export default MovieForm;
