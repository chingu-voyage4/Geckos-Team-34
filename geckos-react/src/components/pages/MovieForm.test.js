import React from 'react';
import { shallow, mount } from 'enzyme';

import MovieForm from '../forms/MovieForm';
import Embed from 'semantic-ui-react';

describe('<MovieForm />', () => {
  let wrapper;
  const props =  () => {};
  beforeEach(() => {
    wrapper = shallow(<MovieForm submit={ props } />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.length).toBe(1);
  });

  it('should render a form', () => {
    const elem = wrapper.find('Form');
    expect(elem.length).toBe(1);
  });

  it('should render a title input', () => {
    const elem = wrapper.find('#title');

    elem.simulate('change', { target: { name: 'title', value: 'title' } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('title')).toEqual('title');
  });

  it('should render a release date input and update the components state.releaseDate value onChange', () => {
    const elem = wrapper.find('#releaseDate');

    elem.simulate('change', { target: { name: 'releaseDate', value: 2018 } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('releaseDate')).toEqual(2018);
  });

  it('should render a run time input and update the components state.runTime value onChange', () => {
    const elem = wrapper.find('#runTime');

    elem.simulate('change', { target: { name: 'runTime', value: 123 } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('runTime')).toEqual(123);
  });

  it('should render a type radio input and update the components state.type value onChange', () => {
    const elemMovie = wrapper.find('[value="movie"]');
    const elemTV = wrapper.find('[value="tv"]');


    elemMovie.simulate('change', {
      target: { name: 'type', checked: true }
    });

    expect(elemMovie.length).toBe(1);
    expect(elemTV.length).toBe(1);
    expect(wrapper.state('type')).toEqual('movie');
  });

  it('should render a star rating input', () => {
    const elem = wrapper.find('#rating');

  //  elem.simulate('select', { name: 'rating', target: { value: 0 } });
    elem.simulate('rate', { target: { value: 1 } });


    expect(wrapper.state('rating')).toEqual(1);
    expect(elem.length).toBe(1);
  });

  it('should render a genre input and update the components state.genre value onChange', () => {
    const elem = wrapper.find('#genre');

    elem.simulate('change', { target: { name: 'genre', value: 'action' } });

    expect(wrapper.state('genre')).toEqual('action');
    expect(elem.length).toBe(1);
  });

  it('should render a director text input and update the components state.director value onChange', () => {
    const elem = wrapper.find('#director');

    elem.simulate('change', { target: { name: 'director', value: 'director' } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('director')).toEqual('director');
  });

  it('should render a producers text input and update the components state.producers value onChange', () => {
    const elem = wrapper.find('#producers');

    elem.simulate('change', { target: { name: 'producers', value: 'producers' } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('producers')).toEqual('producers');
  });

  it('should render a plot description textarea', () => {
    const elem = wrapper.find('#plot');
    expect(elem.length).toBe(1);
  });

  it('should render a language dropdown input and update the components state.language value onChange', () => {
    const elem = wrapper.find('#language');

    elem.simulate('change', { target: { name: 'language', value: 'english' } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('language')).toEqual('english');
  });

  it('should render a movie rating input and update the components state.movieRating value onChange', () => {
    const elem = wrapper.find('#movieRating');

    elem.simulate('change', { target: { name: 'movieRating', value: 'g' } });

    expect(elem.length).toBe(1);
    expect(wrapper.state('movieRating')).toEqual('g');
  });

});
