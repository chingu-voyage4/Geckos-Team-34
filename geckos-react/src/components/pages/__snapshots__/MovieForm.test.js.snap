// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieForm /> should render correctly 1`] = `
<Form
  as="form"
  onSubmit={[Function]}
>
  <FormGroup>
    <FormField
      width={4}
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Mean Girls"
        type="text"
        value=""
      />
    </FormField>
    <FormField
      width={2}
    >
      <label
        htmlFor="releaseDate"
      >
        Release Date
      </label>
      <input
        id="releaseDate"
        maxLength={4}
        name="releaseDate"
        onChange={[Function]}
        placeholder="2018"
        type="text"
        value=""
      />
    </FormField>
    <FormField
      width={2}
    >
      <label
        htmlFor="runTime"
      >
        Run Time(mins)
      </label>
      <input
        id="runTime"
        maxLength={3}
        name="runTime"
        onChange={[Function]}
        placeholder="120"
        type="text"
      />
    </FormField>
  </FormGroup>
  <FormGroup
    inline={true}
  >
    <label>
      Type
    </label>
    <FormRadio
      as={[Function]}
      checked={false}
      control={[Function]}
      label="Movie"
      name="type"
      onChange={[Function]}
      value="movie"
    />
    <FormRadio
      as={[Function]}
      checked={false}
      control={[Function]}
      label="TV"
      name="type"
      onChange={[Function]}
      value="tv"
    />
  </FormGroup>
  <FormGroup
    inline={true}
  >
    <label
      htmlFor="rating"
    >
      Your rating
    </label>
    <Rating
      clearable="auto"
      defaultRating={1}
      icon="star"
      id="rating"
      maxRating={5}
      name="rating"
      onRate={[Function]}
    />
  </FormGroup>
  <FormGroup>
    <FormField
      width={2}
    >
      <label
        htmlFor="genre"
      >
        Genre
      </label>
      <input
        id="genre"
        name="genre"
        onChange={[Function]}
        placeholder="Action"
        type="text"
        value=""
      />
    </FormField>
    <FormField
      width={4}
    >
      <label
        htmlFor="director"
      >
        Director
      </label>
      <input
        id="director"
        name="director"
        onChange={[Function]}
        placeholder="Some Director"
        type="text"
        value=""
      />
    </FormField>
    <FormField
      width={4}
    >
      <label
        htmlFor="producers"
      >
        Producers
      </label>
      <input
        id="producers"
        name="producers"
        onChange={[Function]}
        placeholder="Some producer"
        type="text"
        value=""
      />
    </FormField>
  </FormGroup>
  <FormField
    width={8}
  >
    <label
      htmlFor="plot"
    >
      Plot
    </label>
    <textarea
      cols="30"
      id="plot"
      name="plot"
      onChange={[Function]}
      rows="10"
      value=""
    />
  </FormField>
  <FormGroup>
    <FormField
      width={2}
    >
      <label
        htmlFor="language"
      >
        Language
      </label>
      <select
        id="language"
        name="language"
        onChange={[Function]}
        value=""
      >
        <option
          value="english"
        >
          English
        </option>
        <option
          value="spanish"
        >
          Spanish
        </option>
        <option
          value="french"
        >
          French
        </option>
        <option
          value="chinese"
        >
          Chinese
        </option>
      </select>
    </FormField>
    <FormField
      width={2}
    >
      <label
        htmlFor="movieRating"
      >
        Rating
      </label>
      <select
        id="movieRating"
        name="movieRating"
        onChange={[Function]}
        value=""
      >
        <option
          value="g"
        >
          G
        </option>
        <option
          value="pg"
        >
          PG
        </option>
        <option
          value="pg-13"
        >
          PG-13
        </option>
        <option
          value="r"
        >
          R
        </option>
      </select>
    </FormField>
  </FormGroup>
  <ImageUpload
    imgFile={[Function]}
  />
  <Button
    as="button"
    primary={true}
  >
    Add Movie
  </Button>
</Form>
`;
